src/Gaffer/Spreadsheet.cpp

```diff
@@ -173,13 +179,206 @@ class RowsMapScope : boost::noncopyable, public Context::SubstitutionProvider
 
 GAFFER_PLUG_DEFINE_TYPE( Spreadsheet::RowsPlug );
 
+// Manages a data structure mapping between row name
+// and row, so we can implement `RowsPlug::row()` without
+// linear search.
+class Spreadsheet::RowsPlug::RowNameMap
+{
+
+	public :
+
+		RowNameMap( Spreadsheet::RowsPlug *rowsPlug )
+			:	m_rowsPlug( rowsPlug )
+		{
+			rowsPlug->childAddedSignal().connect( boost::bind( &RowNameMap::childAdded, this, ::_2 ) );
+			rowsPlug->childRemovedSignal().connect( boost::bind( &RowNameMap::childRemoved, this, ::_2 ) );
+			rowsPlug->parentChangedSignal().connect( boost::bind( &RowNameMap::parentChanged, this ) );
+		}
+
+		RowPlug *row( const string &rowName )
+		{
+			if( !m_plugSetConnection.connected() )
+			{
+				// If we can't track name changes, we must do
+				// a brute force update. This should basically never
+				// happen, so we put no effort into making something more
+				// performant.
+				updateRowNames();
+			}
+
+			const auto &index = m_map.get<1>();
+			auto range = index.equal_range( rowName );
+			if( range.first == index.end() )
+			{
+				return nullptr;
+			}
+			else if( range.first == range.second )
+			{
+				// Single row with the requested name.
+				// This is the common case we are optimising
+				// for.
+				return range.first->plug;
+			}
+			else
+			{
+				// Multiple rows with the requested name.
+				// We must return the first one.
+				size_t minIndex = std::numeric_limits<size_t>::max();
+				RowPlug *result = nullptr;
+				for( auto it = range.first; it != range.second; ++it )
+				{
+					/// \todo We need GraphComponent to provide constant-time
+					/// access to a child's index.
```

Comment:
> we find_if once, and pick the first one thats in range

I think it's the same complexity either way isn't it? Either it's `numChildren` searches in `range`, or it's `rangeSize` searches in `numChildren`? I guess the `find_if()` can terminate earlier, so will average half the number of operations?

The main reason I structured it this way was so that it was ready for a call to a constant-time `GraphComponent::index()` method to be slotted in when it was available. I'm not overly concerned about performance for this edge case in the short term, because the builders of large spreadsheets are going to be ensuring that `len( range ) == 1`.